### PR TITLE
Feature request- Bully tool

### PR DESCRIPTION
**Pull request by chandelshubham63**
Implemented the Bully tool in DDT. This tool is used for brute-forcing WPS PIN authentication and can be used to test the security of Wi-Fi networks. The tool allows the user to specify the access point interface or SSID, along with a password list and an optional BSSID. The tool then attempts to brute-force the WPS PIN using the provided settings and outputs the results to the console.

Changes:
Added a new component for the Bully tool in src/components/Bully.tsx
Added the Bully tool to the ROUTES array in src/components/RouteWrapper.tsx
Added the Bully tool to the tauri.conf.json file to allow it to be executed as a command.

Tested the tool with various settings and confirmed that it is working as expected.
To test the tool, simply navigate to the Bully tool in DDT and provide the necessary settings. The tool will then attempt to brute-force the WPS PIN and output the results to the console.

### CODE REVIEW

File: src/components/Bully/Bully.tsx [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    "Bully is a tool for brute-forcing WPS PIN authentication.\n" +
+    "To use the Bully tool for WPS attacks, follow these steps:\n" +
+    "- Enter the access point (AP) interface: Specify the interface name (e.g., wlan0) of the target Wi-Fi network you want to attack.\n" +
+    "- Provide the password list: Input the path to the file containing the list of passwords to be tried during the brute-force attack.\n" +
+    "- Optionally, provide the MAC address (BSSID) or Extended SSID (ESSID) of the target access point.\n" +
+    "- Initiate the brute-force attack: Click the 'Start Bully Attack' button to begin the attack using the provided options.\n" +
+    "- Review the output: After the attack is complete, review the output to identify the cracked WPS PIN, if successful.\n";
+
+interface FormValuesType {
+    Interface: string;
```
Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:20]: - [ ] camelCase

File: src/components/Bully/Bully.tsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    PasswordList: string;
```
Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:21]: - [ ] camelCase

File: src/components/Bully/Bully.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    BSSID: string;
```
Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:22]: - [ ] camelCase

File: src/components/Bully/Bully.tsx [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    ESSID: string;
```
Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:23]: - [ ] camelCase

File: src/components/Bully/Bully.tsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+const Bully = () => {
```
Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:26]: - [ ] Comment to denote state variables.

Check Aircrack for review.

File: src/components/Bully/Bully.tsx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        initialValues: {
+            Interface: "",
+            PasswordList: "",
+            BSSID: "",
+            ESSID: "",
+        },
+    });
+
+    // Function to handle form submission
+    const onSubmit = async (values: FormValuesType) => {
```
Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:43]: - [ ] In line comments missing.

File: src/components/Bully/Bully.tsx [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+            args.push("-b", values.BSSID);
+        }
+
+        if (values.ESSID) {
+            args.push("-e", values.ESSID);
+        }
+
+        // Execute Bully command with constructed arguments
+        try {
+            const commandOutput = await CommandHelper.runCommand("bully", args);
```
Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:66]: - [ ] Need to use getpidandoutput this is deprecated command

Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:66]: Need to use canceloverlay

File: src/components/Bully/Bully.tsx [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    const clearOutput = () => {
+        setOutput("");
+        setHasSaved(false);
+        setAllowSave(false);
+    };
+
+    // Render component
+    return (
+        <form onSubmit={form.onSubmit(onSubmit)}>
+            <LoadingOverlay visible={loading} />
```
Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:94]: - [ ] Need to implement the checkinstalled

See Aircrack for example.

**Review by bradleybogg — CHANGES_REQUESTED** [review]
Good Work!

please see requested changes before proceeding to application review.

File: src/components/Bully/Bully.tsx [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+            // Now that loading has completed, allow the user to save the output to a file.
+            setAllowSave(true);
+            setHasSaved(false);
+        },
+        [handleProcessData] // Dependency on the handleProcessData callback
+    );
+
+    // Actions taken after saving the output
+    const handleSaveComplete = () => {
```
Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:122]: - [ ] Upgrade to docstring.

File: src/components/Bully/Bully.tsx [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+                    <TextInput label="Access Point Interface" required {...form.getInputProps("interface")} />
+                    <TextInput label="Password List" required {...form.getInputProps("passwordList")} />
+                    {advancedMode && (
+                        <>
+                            <TextInput label="MAC Address (BSSID)" {...form.getInputProps("bssid")} />
+                            <TextInput label="Extended SSID (ESSID)" {...form.getInputProps("essid")} />
+                        </>
+                    )}
+                    <Button type="submit" disabled={loading}>
+                        Start Bully Attack
```
Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:215]: <Button type={"submit"}>Start {title}</Button>

- [ ] Use now known variables.

File: src/components/Bully/Bully.tsx [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+                    setOpened={setOpened}
+                    feature_description={description_userguide}
+                    dependencies={dependencies}
+                ></InstallationModal>
+            )}
+
+            <form onSubmit={form.onSubmit(onSubmit)}>
+                {LoadingOverlayAndCancelButton(loading, pid)}
+                <Stack>
+                    {UserGuide(title, description_userguide)}
```
Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:199]: - [ ] camelCase

File: src/components/Bully/Bully.tsx [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+        setAllowSave(false);
+    }, [setOutput]);
+
+    return (
+        <>
+            {!loadingModal && (
+                <InstallationModal
+                    isOpen={opened}
+                    setOpened={setOpened}
+                    feature_description={description_userguide}
```
Comment (bradleybogg) [inline src/components/Bully/Bully.tsx:191]: - [ ] camelCase

**Review by bradleybogg — CHANGES_REQUESTED** [review]
Overall @chandelshubham63 

Very good work. I appreciate that you have implemented the newer versions of the codebase. There is some more work to be done to ensure that this component is also 'baselined'.

Cheers

**Comment by bradleybogg** [comment]
@chandelshubham63 

I have rectified your errors because you did not push the libraries you were using from main into your branch.
I cannot complete this application testing without the wireless adapter. Please upload evidence of the tool working as expected.

Thank you!

**Review by bradleybogg — COMMENTED** [review]
Please upload as requested.

**Comment by zkkok** [comment]
Hi @chandelshubham63 ,
Please resolve the conflicts before application testing

**Review by zkkok — CHANGES_REQUESTED** [review]
Hi @chandelshubham63,

Please review your code again, the code doesn't run properly. `-i` is not use to specify the interface and `-f` is not an available parameter for the tool please have a look at the kali documentation for the tool.
https://www.kali.org/tools/bully/

**Comment by chandelshubham63** [comment]
Thanks for the feedback @zkkok . I modified the code to follow the requested changes

**Comment by zkkok** [comment]
@chandelshubham63 ,Application testing has been done.

The tools runs smoothly and provided the expected outputs.

Input:
image

Output:
image

**Review by zkkok — APPROVED** [review]

**Pull request merged**